Refactor volume controller parameters into a structure

jsafrane opened the pull request:
`persistentvolumecontroller.NewPersistentVolumeController` has 11 arguments now,
put them into a structure.

Also, rename `NewPersistentVolumeController` to `NewController`, `persistentvolume`
is already name of the package.

Fixes #30219

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32129)

<!-- Reviewable:end -->

jdef commented:
lgtm

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/32129#-:-KR1OgqwG9Up-0gG8tOg:bb7k0qh)*
<!-- Sent from Reviewable.io -->

jsafrane commented:
rebased

jdef commented:
Reviewed 5 of 5 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/32129)*
<!-- Sent from Reviewable.io -->

jdef commented:
@mikedanese PTAL

jsafrane commented:
rebased

jsafrane commented:
rebased again

jsafrane commented:
rebased, @mikedanese, PTAL

jsafrane commented:
rebased (this gets pretty annoying...)

jsafrane commented:
@k8s-bot test this issue: 32685

jsafrane committed:
Refactor volume controller parameters into a structure

persistentvolumecontroller.NewPersistentVolumeController has 11 arguments now,
put them into a structure.

Also, rename NewPersistentVolumeController to NewController, persistentvolume
is already name of the package.

Fixes #30219

mikedanese commented:
Sorry for the delay, I suggest also creating a substruct in this struct as well for parameters specific to pvc controller:

https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/componentconfig/types.go#L493

mikedanese commented on test/integration/persistentvolumes/persistent_volumes_test.go:
```diff
@@ -1124,20 +1124,14 @@ func createClients(ns *api.Namespace, t *testing.T, s *httptest.Server, syncPeri
-		nil,  // classSource
-		nil,  // eventRecorder
-		true) // enableDynamicProvisioning
+	params := persistentvolumecontroller.ControllerParameters{
```
Why don't you inline this?

jsafrane replied:
inlined

mikedanese commented:
basically lgtm. one comment, then I will apply label.

mikedanese commented:
Feel free to apply yourself.

jsafrane commented:
>I suggest also creating a substruct in this struct as well for parameters specific to pvc controller:
> https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/componentconfig/types.go#L493

The link points inside `LeaderElectionConfiguration`. Looking at the sources, leader is elected at controller-manager level and individual controllers should not mess up with it.

So, it must have been link to `KubeControllerManagerConfiguration`. There already is `VolumeConfiguration` (shared with attach/detach controller) and `PVClaimBinderSyncPeriod` . That's IMO not worth a sub-structure. I'll add a substructure when the number of PV controller related items grows.

jsafrane commented:
@k8s-bot unit test this issue: #33558

jsafrane commented:
@k8s-bot gci test this

jsafrane commented:
@k8s-bot gci gce e2e test this
